### PR TITLE
Kp/bst traversal Review Third

### PR DESCRIPTION
**Pull request by kp483**
Before: 
- Serialization lesson from C16
- instructor class lesson overview doc from C16

Now:
- Revises C16 serialization for clarity
- Adds BFS pseudocode (class livecode will be BFS walkthrough)
- Adds explanation of Big O for serialization algorithms
- Revises instructor class lesson overview to be up to date with C17 lesson plan
- Updates `TreeNode` class in BST implementation lesson to use `not` for comparison instead of `== None` for code style conventions

For review 
- Clarity of explanation
- Flow and structure
-Lines 267 - 322 of the travesal file are unfinished - should we ask students to implement a `height` function recursively?
    - If finished they would be asked to complete a height exercise recursively
    - If removed, we would instead provide pseudocode or an example implementation of height

### CODE REVIEW

File: 03-Binary-Search-Trees/04-Binary-Search-Trees-Traversal.md [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
+A _traversal_ is an action visiting each node in a data structure such as a tree.  There are several kinds of traversals, Breadth First Traversals which visit each node level, by level and Depth First Traversals which visit a node's children before it's siblings.
+
 
 ### Depth First Traversals
 
-Unlike linear data structures like arrays or linked list which have only one logical way to traverse them, trees can be traversed in different ways.  In a _Depth-first traversal_ you explore the children and grandchildren of a node before moving to it's sibling.
+Unlike linear data structures like arrays or linked list which have only one logical way to traverse them, trees can be traversed in different ways.  In a _depth first traversal_ you explore the children, grandchildren, and any further descendants of a node before moving to its sibling and traversing the sibling's descendants.
 
 There are three standard types of depth-first traversals:
```
Comment (char-adadev) [inline 03-Binary-Search-Trees/04-Binary-Search-Trees-Traversal.md:12]: I had a professor who taught me this method for pre-order, post-order, and in-order traversals that has stuck with me over the years. I'm not sure if it's worth adding here but I wanted to mention it: https://blog.devgenius.io/dfs-depth-first-search-traversal-techniques-short-and-sweet-1e4c134babcf

I'm also happy to demonstrate this method in class!

**Review by char-adadev — COMMENTED** [review]
Thank you for working on this! As for your question about adding the recursive height exercise to this activity, I think it would be a good addition here, if time allows us to add it. Otherwise, maybe we can add the pseudo code and have it be a classroom activity/optional practice? Lemme know and I can approve as soon as we have come to a consensus.

File: 03-Binary-Search-Trees/04-Binary-Search-Trees-Traversal.md (hunk 1 above)
Comment (kp483) [reply 03-Binary-Search-Trees/04-Binary-Search-Trees-Traversal.md:12]: Added as a resource! Can definitely go over in class if time

**Comment by kp483** [comment]
@char-adadev Added the heigh method. Can you re-review?

**Review by char-adadev — APPROVED** [review]
Great work here! I had a small suggestion around the naming of the sections:

The first two sections are named "Binary Search Trees". I would suggest renaming the first one to "Review of Ordered Collections of Data" or something similar. Approving for now as that's a quick follow-up and is inconsequential to the content of the lesson.

**Pull request merged**